Add similarity functions to calculate similarity of given vectors

QuerthDP opened the pull request:
### Motivation

It should be possible to return the similarity of vectors in CQL statements following the [Cassandra compatible syntax](https://cassandra.apache.org/doc/latest/cassandra/getting-started/vector-search-quickstart.html#query-vector-data-with-cql):

```
SELECT comment, similarity_cosine(comment_vector, [0.1, 0.15, 0.3, 0.12, 0.05])
    FROM cycling.comments_vs;
```

Although the calculations are slow, and we already have calculated results returned via Vector Store API,
we need the functionality as it allows us to calculate similarity of vectors not stored in vector indexes.

It will be needed for [quantization and rescoring](https://scylladb.atlassian.net/wiki/spaces/RND/pages/195985800/Quantization+and+Rescoring).

The feature is also a nice-to-have in testing as requested many times by testing and CX teams.

The optimized version utilizing already calculated distances from Vector Store without a need of rescoring will be coming soon after via https://github.com/scylladb/scylladb/pull/27991.

---

### Solution

The patch adds functions:
- `similarity_cosine(<vector>, <vector>)`,
- `similarity_euclidean(<vector>, <vector>)`,
- `similarity_dot_product(<vector>, <vector>)`

Where `<vector>` is either a column of type `VECTOR<FLOAT, N>` or a vector of floats literal.

These functions can be called with every `SELECT` query, not only ANN vector queries as opposed to https://github.com/scylladb/scylladb/pull/25993.

The similarity calculations are implemented inspired by [USearch's implementation](
https://github.com/unum-cloud/USearch/blob/a2f17599101729d667dc0260dd278852d9098183/include/usearch/index_plugins.hpp#L1304-L1385) and made compatible with [Cassandra's documentation](https://cassandra.apache.org/doc/5.0/cassandra/developing/cql/functions.html#vector-similarity-functions).
That would guarantee the results in ScyllaDB are calculated using the exact same algorithms as used in Vector Store indexes.

---

Fixes: SCYLLADB-88
Fixes: SCYLLADB-89

New feature, should land into 2026.1

QuerthDP commented:
Changelog:
- rename `vector_search_fcts` to `vector_similarity_fcts`
- remove unrelated `vector_search_client` tests
- adjust CQLpy tests
- remove USearch dependency
- implement calculations of similarity metrics
- adjust function arguments validation to accept either vector column or literal as both arguments

QuerthDP commented:
Opening to run the CI.
Do not review yet.

TODO:
- add happy path tests
- update docs
- reword commit messages

QuerthDP commented:
Failed on: https://github.com/scylladb/scylladb/issues/27510, but I believe the Jenkins job was run before the fix.

QuerthDP commented:
Changelog:
- update docs

@swasik please take a look at the documentation, comments would be appreciated

avikivity commented on cql3/Cql.g:
```diff
@@ -2387,6 +2410,10 @@ K_MUTATION_FRAGMENTS:    M U T A T I O N '_' F R A G M E N T S;
+K_SIMILARITY_EUCLIDEAN:     S I M I L A R I T Y '_' E U C L I D E A N;
+K_SIMILARITY_COSINE:        S I M I L A R I T Y '_' C O S I N E;
+K_SIMILARITY_DOT_PRODUCT:   S I M I L A R I T Y '_' D O T '_' P R O D U C T;
+
```
Ah, you explain it in the commit message. It's a pity to add this special case.

Please try to add constants to the general case so we don't uglify the grammar unnecessarily.

nyh replied:
I'm also puzzled why this needs to be a new grammar, it looks just like any other function call, no? But I'll leave this part of the discussion to @avikivity he's the expert in the CQL expression parsing.

avikivity commented on cql3/functions/vector_similarity_fcts.hh:
```diff
@@ -0,0 +1,70 @@
+        : vector_similarity_fct(arg_types, "similarity_dot_product") {
+    }
+
+    virtual bytes_opt execute(std::span<const bytes_opt> parameters) override;
```
This would be avoided by ordering the series bottom-up.

avikivity commented on cql3/functions/functions.cc:
```diff
@@ -413,13 +413,15 @@ functions::get(data_dictionary::database db,
+            fun = ::make_shared<similarity_euclidean_fct>(arg_types);
         } else if (*similarity_fun_it == SIMILARITY_DOT_PRODUCT_FUNCTION_NAME) {
-            fun = ::make_shared<similarity_dot_product_fct>();
+            fun = ::make_shared<similarity_dot_product_fct>(arg_types);
```
This should have been part of the previous patch. It's just confusing to review something broken that is later fixed.

avikivity commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -7,10 +7,60 @@
+        validate_vector_type(type, provided_args[0]);
+        validate_vector_type(type, provided_args[1]);
+        return {type, type};
+    }
```
See 81821d26cd7ff6ad6af88559e899055ad4d27551 for a similar case.

avikivity replied:
Ah, the code takes care of it.

avikivity commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,92 @@
+#
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+
+import pytest
```
> - null values (Scylla only, as Cassandra returns null here)

It's correct to return NULL on NULL input. NULL means UNKNOWN. Comparing a vector with an unknown vector yields an unknown result.

QuerthDP replied:
Changed

avikivity commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,32 @@ A number of functions are provided to “convert” the native types into binary
 ``0x0000000000000003`` and ``blobAsBigint(0x0000000000000003)`` is ``3``.
 
+
+Vector similarity functions :label-caution:`Experimental` :label-note:`ScyllaDB Cloud`
```
The functions are available everywhere (though not very useful without the search component).

swasik commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,32 @@ A number of functions are provided to “convert” the native types into binary
+
+These functions return a ``float`` value representing the similarity distance between the given vectors for each row.
+The lower the distance value, the more similar the vectors are.
+The distance values are computed using the selected similarity metric by invoking its corresponding function.
```
Instead of it we can provide the information that functions does not normalize vectors.

knowack1 commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,37 @@
+}
+
+bytes_opt similarity_cosine_fct::execute(std::span<const bytes_opt> parameters) {
+    return vector_similarity_fct::execute(parameters);
```
I think there is no sense to override `vector_similarity_fct::execute` to call `vector_similarity_fct::execute`.

QuerthDP replied:
Good comment, yet already fixed.

knowack1 commented on cql3/functions/vector_similarity_fcts.hh:
```diff
@@ -0,0 +1,58 @@
+    virtual bytes_opt execute(std::span<const bytes_opt> parameters) override;
+};
+
+class similarity_cosine_fct : public vector_similarity_fct {
```
Consider not creating these derived classes. I would rather keep only vector_similarity_fct and dispatch to the proper similarity function in vector_similarity_fct::execute.

knowack1 commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -61,14 +61,80 @@ std::vector<data_type> vector_similarity_fct::provide_arg_types(
+    float result_if_zero[2][2];
+    result_if_zero[0][0] = 1 - dot_product / (std::sqrt(norm_a) * std::sqrt(norm_b));
+    result_if_zero[0][1] = result_if_zero[1][0] = 1;
+    result_if_zero[1][1] = 0;
```
Is `result_if_zero[1][0]` not needed ?

knowack1 commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -129,3 +130,95 @@ def test_vector_similarity_fails_on_vector_of_different_size(cql, test_keyspace,
+@pytest.mark.parametrize("similarity_function", ["cosine", "euclidean", "dot_product"])
+def test_vector_similarity_with_two_columns(cql, test_keyspace, similarity_function):
+    schema = 'pk int primary key, v1 vector<float, 3>, v2 vector<float, 3>'
+    data = [
```
Please add a test case for zero vectors (e.g., [0,0,0]) to verify that division-by-zero scenarios are handled correctly.

knowack1 commented on docs/cql/functions.rst:
```````````````````````````````````````````````````````````````````````````````````````diff
@@ -227,6 +227,32 @@ A number of functions are provided to “convert” the native types into binary
+Vector similarity functions :label-caution:`Experimental` :label-note:`ScyllaDB Cloud`
+``````````````````````````````````````````````````````````````````````````````````````
+
+To obtain the similarity distances of the given vectors, use a ``SELECT`` query::
```````````````````````````````````````````````````````````````````````````````````````
The phrase 'similarity distances' is ambiguous because 'similarity' and 'distance' are distinct concepts.

QuerthDP replied:
Fixed

QuerthDP commented:
Changelog:
- remove unnecessary `execute` overrides

QuerthDP commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    return sum;
+}
```
Thanks for the comment. It does match both Cassandra and USearch's approach.

QuerthDP commented on cql3/functions/functions.cc:
```diff
@@ -398,6 +402,30 @@ functions::get(data_dictionary::database db,
+            fun = ::make_shared<similarity_dot_product_fct>();
+        }
+
+        validate_types(db, keyspace, schema.get(), fun, provided_args, receiver_ks, receiver_cf);
```
The validation is there

QuerthDP commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -61,14 +61,80 @@ std::vector<data_type> vector_similarity_fct::provide_arg_types(
+    float norm_b = 0.0f;
+
+    for (size_t i = 0; i < v1.size(); ++i) {
+        float a = value_cast<float>(v1[i]);
```
SCYLLADB-178

m-szymon commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -61,14 +61,80 @@ std::vector<data_type> vector_similarity_fct::provide_arg_types(
+    }
+
+    float result_if_zero[2][2];
+    result_if_zero[0][0] = 1 - dot_product / (std::sqrt(norm_a) * std::sqrt(norm_b));
```
It did. But I guess usearch relies on assumption that division by zero, although terrible in performance, is rare. 
Or it filters out earlier 'zero' vectors. 
And it wants to have branchless code to use SIMD. Although I am not sure if this part is easily vectorized - maybe it can vectorize it when they compute multiple similarity values?
But anyway it looks like some very old code, maybe copied from some GPU kernels.
I believe modern compilers and CPUs can use masked division directly from simple if/else clause.

QuerthDP commented on cql3/functions/functions.cc:
```diff
@@ -398,6 +402,30 @@ functions::get(data_dictionary::database db,
+    };
+
+    auto similarity_fun_it = std::find_if(SIMILARITY_FUNCTIONS.begin(), SIMILARITY_FUNCTIONS.end(), [&name](const auto& f) {
+        return name.has_keyspace() ? name == f : name.name == f.name;
```
I don't see those multiple places in this PR. If you refer to those checks all over the `functions.cc` file, I think it's for the other PR, you might file an issue.

QuerthDP commented:
Changelog:
- adjusted docs
- added tests for bind variables
- added tests for zero vectors
- added exception throw on zero vectors to match Cassandra
- refactored the order of commits

TODO:
- fix commit messages

QuerthDP commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -90,3 +91,72 @@ def test_vector_similarity_fails_on_vector_of_different_size(cql, test_keyspace,
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (pk) VALUES (1)")
+        result = cql.execute(f"SELECT pk, similarity_{similarity_function}({v1}, {v2}) FROM {table}")
+
```
Added

avikivity commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,176 @@
+        cql.execute(f"SELECT pk, similarity_{similarity_function}([1.1, 1.2, 20.25], [1.8, 0.5, 20.03]) FROM {table}")
+        cql.execute(f"SELECT pk, similarity_{similarity_function}(v1, v2) FROM {table}")
+        cql.execute(f"SELECT pk, similarity_{similarity_function}(v1, [1.8, 0.5, 20.03]) FROM {table}")
+        cql.execute(f"SELECT pk, similarity_{similarity_function}([1.1, 1.2, 20.25], v2) FROM {table}")
```
Should also test that the results match expectations (and match Cassandra results for the same queries), so the actual calculations don't regress.

nyh commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,64 @@
+    if (norm_a == 0 || norm_b == 0) {
+        throw exceptions::invalid_request_exception("Function system.similarity_cosine doesn't support all-zero vectors");
+    }
+    return (1 + (dot_product / (std::sqrt(norm_a) * std::sqrt(norm_b)))) / 2;
```
I need an explanation why this isn't just `dot_product / (std::sqrt(squared_norm_a) * std::sqrt(squared_norm_b))`. If it's correct, this explanation should be in the commit message.

QuerthDP replied:
https://github.com/scylladb/scylladb/pull/27524#discussion_r2643749669

nyh commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,64 @@
+        sum += diff * diff;
+    }
+
+    return (1 / (1 + sum));
```
I'm also confused whether we're trying to be compatible with Cassandra, with usearch, with both or neither.

QuerthDP replied:
We're trying to be compatible with Cassandra.
USearch was the initial inspiration to calculate **distances** in the identical way as we do in Vector Store.
As the PR's life progressed, we did switch to **similarity** instead and implemented it the way Cassandra returns.

Either way, both of them calculate "Squared Euclidean" as it doesn't change the ordering (but does change the similarity score so I don't know why).

QuerthDP replied:
Their implementation is just imported from jVector.
And jVector calculates the similarity like this:

```
/** Euclidean distance */
  EUCLIDEAN {
    @Override
    public float compare(VectorFloat<?> v1, VectorFloat<?> v2) {
      return 1 / (1 + VectorUtil.squareL2Distance(v1, v2));
    }
  },

  /**
   * Dot product. NOTE: this similarity is intended as an optimized way to perform cosine
   * similarity. In order to use it, all vectors must be normalized, including both document and
   * query vectors. Using dot product with vectors that are not normalized can result in errors or
   * poor search results. Floating point vectors must be normalized to be of unit length, while byte
   * vectors should simply all have the same norm.
   */
  DOT_PRODUCT {
    @Override
    public float compare(VectorFloat<?> v1, VectorFloat<?> v2) {
      return (1 + VectorUtil.dotProduct(v1, v2)) / 2;
    }
  },

  /**
   * Cosine similarity. NOTE: the preferred way to perform cosine similarity is to normalize all
   * vectors to unit length, and instead use {@link VectorSimilarityFunction#DOT_PRODUCT}. You
   * should only use this function if you need to preserve the original vectors and cannot normalize
   * them in advance. The similarity score is normalised to assure it is positive.
   */
  COSINE {
    @Override
    public float compare(VectorFloat<?> v1, VectorFloat<?> v2) {
      return (1 + VectorUtil.cosine(v1, v2)) / 2;
    }
  };
```

nyh commented on cql3/Cql.g:
```diff
@@ -1691,6 +1708,12 @@ allowedFunctionName returns [sstring s]
     | K_COUNT                       { $s = "count"; }
     ;
 
+allowedSimilarityFunctionName returns [sstring s]
```
Do we really need to return the string? Can't we just take the current token (or tolower()) of it?

QuerthDP replied:
I guess we can, but yet still we need to define the `allowedSimilarityFunctionName` to say which names are allowed. I don't think it's much better to get the token and transform it instead of just providing a string (as others functions do, look `allowedFunctionName` just above)

nyh replied:
But allowedFunctionName seems to do exactly what I said:
```
    : f=IDENT                       { $s = $f.text; std::transform(s.begin(), s.end(), s.begin(), ::tolower); }
```
Just instead of IDENT you'll need to do something like `(K_SIMILARITY_COSINE | K _SIMILARITY_EUCLIDEAN | ...)`. My point was just that you won't need to repeat the strings "similarity_cosine" et al. yet again.

QuerthDP replied:
Fixed

QuerthDP commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,37 @@ A number of functions are provided to “convert” the native types into binary
+Both arguments must be of the same dimension.
+
+These functions return a ``float`` value representing the similarity between the given vectors for each row.
+The similarity is a float value telling how similar the vectors are in a range from 0 - not similar at all to 1 - the same.
```
Changed to be more precise

nyh commented on cql3/Cql.g:
```diff
@@ -2387,6 +2410,10 @@ K_MUTATION_FRAGMENTS:    M U T A T I O N '_' F R A G M E N T S;
 
 K_VECTOR_SEARCH_INDEXING: V E C T O R '_' S E A R C H '_' I N D E X I N G;
 
+K_SIMILARITY_EUCLIDEAN:     S I M I L A R I T Y '_' E U C L I D E A N;
```
I think you need to add these new keywords to unreserved_keyword? If not, why not?

nyh replied:
Ok. I'll resolve the discussion.

QuerthDP commented on cql3/functions/vector_similarity_fcts.hh:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    virtual bool is_pure() const override {
+        return false;
```
According to CoPilot's analysis:

> ```SELECT pk, v, similarity_cosine([0, 0, 0], [0, 0, 0]) FROM table```
 This is a query where both arguments are constant literals. When both arguments are constants and `is_pure()` returns true, the optimizer constant-folds the function call into a bare `constant`, which then crashes `selectable_processes_selection()` ("no way to express SELECT constant in the grammar yet").

>Most pure functions are called with column arguments
Functions like `token()`, `toJson()`, `CAST()`, `writetime()` are almost always called with at least one column.
Since they have a non-constant column argument,`all_terminal` is false, so no constant folding happens and the expression stays as a `function_call`

I think I can truly agree to that.

nyh commented on docs/cql/dml/select.rst:
```diff
@@ -266,15 +266,16 @@ the actual number of results returned may be less than the specified limit.
 For example::
 
-    SELECT image_id FROM ImageEmbeddings
+    SELECT image_id, similarity_cosine(embedding, [0.1, 0.2, 0.3, 0.4])
```
Maybe you better have two examples, one which doesn't return the similarity and just sorts by it (this is what the text below says this example does...), and another example where the similarity is returned?

QuerthDP replied:
Changed

QuerthDP commented:
Changelog:
- made vector similarity functions pure functions
- removed the "no way to support SELECT with constant" error since we do since this patch

nyh commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,363 @@
+
+
+def assert_close(actual, expected):
+    assert math.isclose(actual, expected, abs_tol=1e-5)
```
nitpick: there's no real need to have an assert_close function, you can have an isclose() function returning a boolean and then write
```
assert isclose(...)
```
This would be more pytest-like, and will also allow you to do things like `assert isclose(..) and ...` and so on.

nyh commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,363 @@
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}([1.0, 2.0, 3.0], [1.1, 1.2]) FROM {table}")
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}([1.0, 2.0, 3.0], [1.1, 1.2, 20.25, 123.7]) FROM {table}")
```
The one case you didn't check here (I think) is if you have different columns like
```
v1 vector<float, 3>, v2 vector<float, 3>, v3 vector<float,4>
```
It should be possible to compute the similarity of v1 and v2 (right?) but not v1 and v3.

QuerthDP commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (pk, v) VALUES (0, [0.1, 0.2])")
```
Unluckily yes 🙃

QuerthDP commented:
Changelog:
- fix tests according to @nyh's suggestions
- add a test for failing to call similarity on vector columns of different dimension
- rename `norm_a` to `squared_norm_a` to resemble the true

QuerthDP commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,363 @@
+
+
+@pytest.mark.parametrize("similarity_function", ["cosine", "euclidean", "dot_product"])
+def test_vector_similarity_returns_null_on_null_arguments(cql, test_keyspace, similarity_function):
```
I believe I didn't use the fact of any empty table. In fact I tried to have all the tested tables with any rows so that we don't pass if only we do on empty table.

QuerthDP commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,64 @@
+        dot_product += a * b;
+    }
+
+    return ((1 + dot_product) / 2);
```
https://github.com/scylladb/scylladb/pull/27524#discussion_r2643749669

QuerthDP commented:
Changelog:
- fix reference to similarity functions in docs

QuerthDP commented on docs/cql/functions.rst:
```````````````````````````````````````````````````````````````````````````````````````diff
@@ -227,6 +227,32 @@ A number of functions are provided to “convert” the native types into binary
+Vector similarity functions :label-caution:`Experimental` :label-note:`ScyllaDB Cloud`
+``````````````````````````````````````````````````````````````````````````````````````
+
+To obtain the similarity distances of the given vectors, use a ``SELECT`` query::
```````````````````````````````````````````````````````````````````````````````````````
Actually it was crucial for the query to be presented as code block :)
Reverted

nyh commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,133 @@
+namespace {
+
+// You should only use this function if you need to preserve the original vectors and cannot normalize
+// them in advance.
```
I still think a comment should appear somewhere in this file explaining that the bizarre (for me...) choices of scaling and even squaring of the similarities is meant to be compatible with the exact formulas from Cassandra, and that we have tests to confirm this is actually the case.

nyh commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,40 @@ A number of functions are provided to “convert” the native types into binary
+
+    SELECT comment, similarity_cosine(comment_vector, [0.1, 0.15, 0.3, 0.12, 0.05])
+        FROM cycling.comments_vs
+        ORDER BY comment_vector ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]
```
Do you deliberately want to demonstrate these functions just for ANN queries even though if I understand correctly it can also work for ordinary SELECTs?

QuerthDP replied:
Removed the ANN query reference

nyh commented on test/cqlpy/test_vector_similarity.py:
```diff
@@ -0,0 +1,311 @@
+similarity_functions = {"cosine", "euclidean", "dot_product"}
+
+
+@pytest.fixture
```
please use `@pytest.fixture(scope="module")` to make it clearer that this is a file-local fixture.

nyh reviewed (commented):
Thanks. Look better, thanks. I still have a few small comments.
I'm assuming that Avi is reviewing the cql3 expressions part, I don't have any intuition about that.

QuerthDP commented:
Changelog:
- updated comments to mention why the similarity score calculations look this extraordinary
- adjusted `allowedSimilarityFunctionName` to be more concise and not use all the function name string constants
- updated comments explaining why and how we map the distances to similarity score of [0, 1]
- added an error message mentioning the invalid column type in similarity function call
- updated similarity functions docs not to mention ANN queries as the functions may be used anywhere
- updated the note mentioning that functions do not normalize vectors themselves (cosine does)
- adjusted tests to use the precreated `table1` with all necessary columns and example data for calculations
- added a comment on suspiciously not failing test with set as bind variable

QuerthDP commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,40 @@ A number of functions are provided to “convert” the native types into binary
+These functions return a ``float`` value representing the similarity between the given vectors for each row.
+The similarity value is a floating-point number in a range of [0, 1] that describes how similar two vectors are.
+Values closer to 1 indicate greater similarity.
+These functions do not perform vector normalization prior to computing similarity.
```
Adjusted the docs

QuerthDP commented on cql3/functions/vector_similarity_fcts.cc:
```diff
@@ -0,0 +1,133 @@
+
+    // The cosine similarity is in the range [-1, 1].
+    // It's mapped to [0, 1] for consistency with other similarity functions.
+    return (1 + (dot_product / (std::sqrt(squared_norm_a) * std::sqrt(squared_norm_b)))) / 2;
```
Changed. Thank you
